### PR TITLE
Keep track of EQUALS range in Syntax Tree

### PR DESCRIPTION
**Pull request by nojaf**
I'd like to keep track of the range of the equals token in the Syntax tree.
This is useful for Fantomas to reconstruct code comments (trivia) after equal signs.

```fsharp
let foo = // comment
    value
```

The closest element to assign the comment, in this case, is the equal token and it is beneficial to have this information in the Syntax tree.

I have in mind to add it to:
- [x] SynBinding
- [x] SynExpr.LetOrUseBang
- [x] SynExpr.Record
- [x] SynExpr.AnonRecd
- [x] SynPat.Record
- [x] SynEnumCase 
- [x] SynExpr.For
- [x] SynTypeDefn
- [x] SynTypeDefnSig
- [x] NestedModule
- [x] SynMemberDefn.AutoProperty
- [x] SynArgPats.NamePatPairs
- [ ] ...

### CODE REVIEW

**Comment by nojaf** [comment]
@dsyme ready for review.

File: src/fsharp/SyntaxTree.fs [hunk 1]
```diff
@@ -777,8 +778,9 @@ type SynExpr =
         isUse: bool *
         isFromSource: bool *
         pat: SynPat *
+        equalsRange: Range option *
         rhs: SynExpr *
-        andBangs:(DebugPointAtBinding * bool * bool * SynPat * SynExpr * range) list *
+        andBangs: AndBang list *
```
Comment (dsyme) [inline src/fsharp/SyntaxTree.fs:783]: Please use `SynExprAndBang`

File: src/fsharp/SyntaxTree.fs [hunk 2]
```diff
@@ -940,6 +942,25 @@ type SynExpr =
...
+        debugPoint: DebugPointAtBinding *
+        isUse: bool *
+        isFromSource: bool *
+        pat: SynPat *
+        equalsRange: Range *
+        body: SynExpr *
+        range: Range
+
+[<NoEquality; NoComparison>]
+type RecordInstanceField =
```
Comment (dsyme) [inline src/fsharp/SyntaxTree.fs:957]: Please use `SynExprRecordField`

File: src/fsharp/SyntaxTreeOps.fs [hunk 3]
```diff
@@ -618,7 +618,7 @@ let inferredTyparDecls = SynValTyparDecls(None, true)
 let noInferredTypars = SynValTyparDecls(None, false)
 
 let rec synExprContainsError inpExpr =
-    let rec walkBind (SynBinding(_, _, _, _, _, _, _, _, _, synExpr, _, _)) = walkExpr synExpr
+    let rec walkBind (SynBinding(expr = synExpr)) = walkExpr synExpr
```
Comment (dsyme) [inline src/fsharp/SyntaxTreeOps.fs:621]: Nit: Just `expr=synExpr` (no spaces) thanks

File: src/fsharp/service/FSharpParseFileResults.fs [hunk 4]
```diff
@@ -710,7 +710,7 @@ type FSharpParseFileResults(diagnostics: FSharpDiagnostic[], input: ParsedInput,
                 if not (rangeContainsPos memb.Range pos) then [] else
                 [ match memb with
                   | SynMemberDefn.LetBindings(binds, _, _, _) -> yield! walkBinds binds
-                  | SynMemberDefn.AutoProperty(_attribs, _isStatic, _id, _tyOpt, _propKind, _, _xmlDoc, _access, synExpr, _, _) -> yield! walkExpr true synExpr
+                  | SynMemberDefn.AutoProperty(_attribs, _isStatic, _id, _tyOpt, _propKind, _, _xmlDoc, _access, _, synExpr, _, _) -> yield! walkExpr true synExpr
```
Comment (dsyme) [inline src/fsharp/service/FSharpParseFileResults.fs:713]: Could simplify to `SynMemberDefn.AutoProperty(expr=synExpr)`

File: src/fsharp/service/ServiceInterfaceStubGenerator.fs [hunk 5]
```diff
@@ -515,13 +515,13 @@ module InterfaceStubGenerator =
...
     let internal (|MemberNameAndRange|_|) = function
         | SynBinding(_access, _bindingKind, _isInline, _isMutable, _attrs, _xmldoc, SynValData(Some mf, _, _), LongIdentPattern(name, range), 
-                     _retTy, _expr, _bindingRange, _seqPoint) when mf.MemberKind = SynMemberKind.PropertyGet ->
+                     _retTy, _mEquals, _expr, _bindingRange, _seqPoint) when mf.MemberKind = SynMemberKind.PropertyGet ->
             if name.StartsWithOrdinal("get_") then Some(name, range) else Some("get_" + name, range)
         | SynBinding(_access, _bindingKind, _isInline, _isMutable, _attrs, _xmldoc, SynValData(Some mf, _, _), LongIdentPattern(name, range), 
-                     _retTy, _expr, _bindingRange, _seqPoint) when mf.MemberKind = SynMemberKind.PropertySet ->
+                     _retTy, _mEquals, _expr, _bindingRange, _seqPoint) when mf.MemberKind = SynMemberKind.PropertySet ->
             if name.StartsWithOrdinal("set_") then Some(name, range) else Some("set_" + name, range)
         | SynBinding(_access, _bindingKind, _isInline, _isMutable, _attrs, _xmldoc, _valData, LongIdentPattern(name, range),
```
Comment (dsyme) [inline src/fsharp/service/ServiceInterfaceStubGenerator.fs:523]: Could remove all the `_` things and simplify to one selector for the pattern

File: src/fsharp/service/ServiceInterfaceStubGenerator.fs [hunk 6]
```diff
@@ -730,7 +730,7 @@ module InterfaceStubGenerator =
                 match memberDefn with
                 | SynMemberDefn.AbstractSlot(_synValSig, _memberFlags, _range) ->
                     None
-                | SynMemberDefn.AutoProperty(_attributes, _isStatic, _id, _type, _memberKind, _memberFlags, _xmlDoc, _access, expr, _r1, _r2) ->
+                | SynMemberDefn.AutoProperty(_attributes, _isStatic, _id, _type, _memberKind, _memberFlags, _xmlDoc, _access, _, expr, _r1, _r2) ->
```
Comment (dsyme) [inline src/fsharp/service/ServiceInterfaceStubGenerator.fs:733]: Simplify to one selector

File: src/fsharp/service/ServiceInterfaceStubGenerator.fs [hunk 7]
```diff
@@ -900,10 +900,10 @@ module InterfaceStubGenerator =
                 | SynExpr.DoBang (synExpr, _range) -> 
                     walkExpr synExpr
 
-                | SynExpr.LetOrUseBang (_sequencePointInfoForBinding, _, _, _synPat, synExpr1, synExprAndBangs, synExpr2, _range) -> 
+                | SynExpr.LetOrUseBang (_sequencePointInfoForBinding, _, _, _synPat, _, synExpr1, synExprAndBangs, synExpr2, _range) -> 
                     [
                         yield synExpr1
-                        for _,_,_,_,eAndBang,_ in synExprAndBangs do
+                        for AndBang(body = eAndBang) in synExprAndBangs do
```
Comment (dsyme) [inline src/fsharp/service/ServiceInterfaceStubGenerator.fs:906]: No spaces around `=`

**Review by dsyme — CHANGES_REQUESTED** [review]
Just a few minor things - perhaps go over the diff and apply consistently throughout, e.g. I think we almost always use `id=id` in pattern case selectors

**Comment by nojaf** [comment]
@dsyme I've addressed all the feedback.

**Review by dsyme — APPROVED** [review]

**Comment by dsyme** [comment]
This is great work!

**Pull request merged**